README.md
# abn devel project (PRIVATE)


## structure
* abn/ contains source code
* Use testthat for tests

## Modifications
* bug correction for simulation and prediction
* plot function with mb, fitabn output
* implementation of forumla in: Buildscore/fitabn (modification of rd file) 
* plot function + rd file
* prediction function + rd file
* simulation function + rd file

## To do:
### R Package (release)
* rd file: plot, simulation, prediction
* implementation of test for plot/prediction/simulation functions
* diagnostic plots
* bootstrapping
* Implement parallel computing in fitabn(), buildscorecache() 

### R package (extension):
* Implement new prior MP (multinomial + data separation) 
* Data separation 
* Random effect (already one layer) 
* Multinomial
* Negative binomial, ... 
* Time/Memory resources optimization
* Handling missing data 

ChangeLog
---------------------------
-- ChangeLog - abn --
	---------------------------
2023-05-06 Matteo Delucchi (version 2.7-5)
  * Changed contributors
  * Fixed unicode-character issue in NAMESPACE

2023-01-21 Reinhard Furrer  (version 2.7-4)

  * Changes for JSS article.
  * Minor internal improvements (eliminated tests based on `class()`)
	
	
2022-06-15 Reinhard Furrer  (version 2.7-3, 2.7-1)

  * Fixed a bug in `buildScoreCache.mle()` ()
  * Eliminated a couple potential warnings.
  * Improvement of version `abn.Version()`
  * Cleaned control arguments
  * Improved a few man pages


2022-04-02 Reinhard Furrer  (version 2.7-0)

  * plotAbn(): new arguments, now fully based on class `graphAM`, proper handling of lty, lwd etc.
  * options of `fitAbn()` and `buildScoreCache()` are now handled with a list passed to argument `control=`. Some of the default values have been slightly adapted.
  * Default list available via `fit.control()` and `build.control()`, including more options!
  * Significantly improved help of several functions and examples.
  * Improved display of print methods.
  * Possiblity to use single threaded with `for()` loop and with `foreach()`.
  * Due to proper method handling, the arguments `create.graph` in `searchHillClimber()` and `fitAbn(method="bayes")` has been removed.
  * Argument `verbose` properly implemented (including passing to JAGS functions).
  * New function `compareEG()` to compare two essential graphs.
  * Additional argument checking an cleaning of many error messages.


  * Internal changes:
	Option `error.verbose` properly propagated to C functions.
	New option `trace` passed to "lbfgsb".
	Less ARMA messages, using now `#define ARMA_WARN_LEVEL 1`.
	Some GSL error now caught at R level.
	Severe R and moderate C code cleaning.



2021-06-08 Gilles Kratzer, Reinhard Furrer & Kalina Cherneva (version 2.6.0)
	   * Bug fixes:
	      . initial values for several chains in `simulateAbn()`
	      . new default value for `method` in `fitAbn()`
	      . new default value `method='bic'` in `mostProbable()`
	      . minor bug fixes in `plotAbn()`
	      . minor bug fix in `help(plotAbn)`
	      . number of parent specification for `buildScoreCache( , method='mle')`
	   * Renaming many function arguments for consistency
	   * Reordering, purging and adding function argments
	   * Parallelization of `buildScoreCache(, method='mle')`
	   * Cleaning of RCpp with improved default parameters (speed up 5%)
	   * Adapting authorship of package
	   * Minor cleaning of `Examples` in the help files


2021-04-23 Gilles Kratzer & Reinhard Furrer (version 2.5.0)
	   * Renaming many functions to CamelCase for consistency
	   * Renaming many arguments for consistency
	   * Reordering, purging and adding function argments
	   * Major changes in `plotAbn()`
	   * Cleaning of several help files

2020-10-31 Gilles Kratzer & Reinhard Furrer (version 2.4.0)
	   * Update on citations
	   * Update on ORCID number

2020-10-22 Gilles Kratzer & Reinhard Furrer (version 2.3.0)
	   * Requirement for use of R>=4.0.0

2020-06-13 Gilles Kratzer & Reinhard Furrer (version 2.2.1)
	   * Patch for S4 exports not stated in the package
	   * Testing automatic (testhat.R in tests repository)
	   * Updated version of the vignette

2019-11-22 Gilles Kratzer & Reinhard Furrer (version 2.2)
	   * New vignette describing new functionalities!
	   * Old vignette is store in the﻿ directories inst/bootstrapping_example and inst/old_vignette
	   * Dataset: FCV
	   * Methods related to the class abnFit: family, logLik, AIC, BIC, nobs, coef
	   * Bugs corrected in plotabn()
	   * Compilation warnings corrected
	   * Typos in man help files corrected
	   * Compiler flags that were not compatible with the Solaris OS

2019-07-04 Gilles Kratzer & Reinhard Furrer (version 2.1)
	   * Compiler flags that were not compatible with the Solaris system
	   * Note related to INLA

2019-07-01 Gilles Kratzer & Reinhard Furrer (version 2.0)
	   * Implementation of S3 methods (print, summary, plot)
	   * New classes:
	   		* abnDag, abnCache, abnHeuristic, abnMostprobable, abnHillClimber, abnFit
	   * renaming of functions
	   		* link.strength() -> linkStrength()
	   		* search.heuristic() -> searchHeuristic()
	   		* backward compatibility is not ensured for those functions
	   * New tests implemented for internal validation and some backward compatibility
	   * Dependence with Cairo removed
	   * modification of search.hillclimber(): no interactive plotting possible
	   * buildscorecache() becomes a parametrisable wrapper for the bayesian and the mle methods
	   * fitabn() becomes a parametrisable wrapper for the bayesian and the mle methods
	   * Dataset: adg
	   * Multiple Bugs corrected

2018-07-23 Gilles Kratzer (version 1.2)
	   * (many) new functions
	   * buildscorecache.mle()
	   * fitabn.mle()
	   * plotabn()
	   * simulateabn()
	   * expit(), logit(), odds() and or()
	   * mb()
	   * link.strength()
	   * discretization()
	   * essentialGraph()
	   * compareDag()
	   * infoDag()
	   * simulateDag()
	   * search.heuristic()
	   * entropyData()
	   * miData()

	   * modification of existing functions:
	   * mostprobable()

2016-11-09 Gilles Kratzer (version 1.0.2)
	   * Default ban and retain matrix
	   * Automatic testing procedure of the package

2016-01-11 Pittavino Marta (version 1.0)
	   New vignette uploaded.

2015-12-22 Pittavino Marta (version 0.86)
	   INLA routine updated, ERROR and NOTE addressed.

2014-12-22 Pittavino Marta (version 0.85)
	   * Published Version v0.85 on CRAN

2014-12-09 Pittavino Marta (version 0.85)
           * cleanup file to delete additional Makevars added

2014-12-08 Pittavino Marta (version 0.85)
           * FOSS license --as-cran NOTE addressed, eliminated BuildVignettes NO in the DESCRIPTION file

2014-12-06 Pittavino Marta (version 0.84)
           * v0.84 submitted to CRAN

2014-12-05 Pittavino Marta (version 0.84)

R/plot-abn.R
## plot-abn.R --- Author : Gilles Kratzer Last Modified on: 06/12/2016 Last Modified on: 10/03/2017 Last modification: 19.05.2017 Node color list Last mod: 13.06.2017 Arc direction Last mod:
## 18/07/2017
## major rewrite rf 2021-04

# for final submission elimiate and # `print()` lines


plotabn <- function(...) {
    .Deprecated("plotAbn", msg="'plotabn' is deprecated.\n Use 'plotAbn' instead but note that arguments have slightly changed.")
    dots <- list(...)
    if (!is.null(dots$dag.m)) {
        dag <- dots$dag.m
        dots$dag.m <- NULL
        do.call('plotAbn', c(dag, dots))
    } else  plotAbn(...)
}


plotAbn <- function(dag, data.dists=NULL, markov.blanket.node=NULL,
                    fitted.values=NULL, digits=2, edge.strength=NULL, edge.strength.lwd=5,
                    edge.direction="pc", edge.color="black", edge.linetype="solid", edge.arrowsize=0.6, edge.fontsize=node.fontsize,
                    node.fontsize=12, node.fillcolor=c("lightblue","brown3","chartreuse3"),
                    node.fillcolor.list=NULL, node.shape=c("circle","box","ellipse","diamond"),
                    plot=TRUE , ... )       {

    # Actually, the plot argument is wrong! i do not need the adjacency structure only. I need all but the plotting. i.e., all but the rendering of the graph.


    # The following is not relevant. The nodes are calculated via mb. They are not colored.
    #    if(!is.null(markov.blanket.node) & ("multinomial" %in% (data.dists))) warning("Multinomial nodes are excluded from Markov blanket computation.")

    ## for compatibility purpose
    if(inherits(x=dag, what="abnLearned")){
        data.dists <- dag$score.cache$data.dists;
        dag <- dag$dag
    }
  if(inherits(x=dag, what="abnFit")){
    data.dists <- dag$abnDag$data.dists
    dag <- dag$abnDag$dag
  }
  if (is.null(data.dists)) stop("'data.dist' need to be provided.")
    name <- names(data.dists)


    ## dag transformation
    if (!is.null(dag)) {
        if (is.matrix(dag)) {
            ## run a series of checks on the DAG passed
            dag <- abs(dag)
            ## consistency checks
            diag(dag) <- 0
            dag[dag > 0] <- 1

            ## naming
            if (is.null(rownames(dag))) {
                colnames(dag) <- name
                rownames(dag) <- name
            }
            dag <- check.valid.dag(dag=dag, is.ban.matrix=FALSE, group.var=NULL)
        } else {
            if (grepl("~", as.character(dag)[1], fixed=T)) {
                dag <- formula.abn(f=dag, name=name)
                ## run a series of checks on the DAG passed
                dag <- check.valid.dag(dag=dag, is.ban.matrix=FALSE, group.var=NULL)
            }
        }
    } else {
        stop("'dag' specification must either be a matrix or a formula expression")
    }

    # contains Rgraphviz
    if (edge.direction == "undirected") {
        dag=dag + t(dag)
        dag[dag != 0] <- 1     # this should not be necessary!
    }


    ## create an object graph
    am.graph <- new(Class="graphAM", adjMat=dag,
                    edgemode=ifelse(edge.direction=="undirected","undirected","directed"))

    ## ========= SHAPE =========
    ## Shape: plot differential depending on the distribution
    node.shape <- rep(node.shape, 4)
    shape <- rep(node.shape[1], length(data.dists) )
    shape[data.dists == "binomial"] <- node.shape[2]
    shape[data.dists == "poisson"] <- node.shape[3]
    shape[data.dists == "multinomial"] <- node.shape[4]
    names(shape) <- names(data.dists)

    ## ================= NODE FILLED COLOR =================
    ## fill with default value, change if MB or fillcolor.list is requested
    fillcolor <- rep(node.fillcolor[1], length(data.dists))
    names(fillcolor) <- names(data.dists)

    ## =============== MARKOV BLANKET ===============
    ## Markov Blanket: plot the MB of a given node
    if (!is.null(markov.blanket.node)) {
        markov.blanket <- mb( dag, node=markov.blanket.node, data.dists=data.dists)
        fillcolor[ names(data.dists) %in%  markov.blanket]  <- node.fillcolor[3]
        fillcolor[ names(data.dists) %in%  markov.blanket.node]  <- node.fillcolor[2]

    } else    if (!is.null(node.fillcolor.list)) {
        fillcolor[ names(data.dists) %in%  node.fillcolor.list] <- node.fillcolor[2]
    }

    names.edges <- names(Rgraphviz::buildEdgeList(am.graph))

    ## =============== Fitted values ===============
    ## Plot the fitted values in abn as edges label
#    print(names.edges)
    if (!is.null(fitted.values)) {
        space <- "   "
        edge.label <- c()
        for (i in 1:length(fitted.values)) {
            if ((length(fitted.values[[i]]) > 1)& (data.dists[names(fitted.values)[i]] != "gaussian")) {
                for (j in 1:(length(fitted.values[[i]]) - 1))
                    edge.label <- c(edge.label, paste(space, signif(fitted.values[[i]][j + 1], digits=digits)))
            } else if ((length(fitted.values[[i]]) > 2)& (data.dists[names(fitted.values)[i]] == "gaussian")){
                for (j in 1:(length(fitted.values[[i]]) - 2))
                    edge.label <- c(edge.label, paste(space, signif(fitted.values[[i]][j + 1], digits=digits)))
            }
        }
    } else  edge.label <- rep(" ", length(names.edges))
    names(edge.label) <- names.edges


    ## =================== Arc Strength ===================
    ## Arc strength: plot the AS of the dag arcs
#    if (is.matrix(edge.strength) & (edge.direction != "undirected")) {
    if (is.matrix(edge.strength)) {
        if (any(edge.strength<0)) stop("'edge.strength' should be positive")
        if (any(edge.strength[dag ==0] >0)) stop("'edge.strength' does not match dag")
        min.as <- min(edge.strength[edge.strength > 0])
        max.as <- max(edge.strength[edge.strength > 0])

        edge.strength.norm <- (edge.strength - min.as)/(max.as - min.as)
        edge.strength.norm[edge.strength.norm < 0] <- 0
        edge.lwd <- list()
        for (i in 1:length(dag[1, ])) {
            for (j in 1:length(dag[1, ])) {
                if (dag[i, j] == 1) {
                    edge.lwd <- cbind(edge.lwd, round(edge.strength.lwd * edge.strength.norm[i, j]) + 1)
                }
            }
        }
    } else {
        edge.lwd <- rep(1, length(names.edges))
    }
    class(edge.lwd) <- "character"
    names(edge.lwd) <- names.edges

    ## ====== Plot ======
    attrs <- list(graph=list(rankdir="BT"),
                  node=list(fontsize=node.fontsize, fixedsize=FALSE),
                  edge=list(arrowsize=edge.arrowsize, color=edge.color, lty=edge.linetype, fontsize=edge.fontsize))
    nodeAttrs <- list(fillcolor=fillcolor, shape=shape)
    edgeAttrs <- list(label=edge.label, lwd=edge.lwd)
#     print(edgeAttrs)
#    if (all(shape %in% c("circle","box","ellipse")))  {
    am.graph <- layoutGraph(am.graph, attrs=attrs, nodeAttrs=nodeAttrs, edgeAttrs=edgeAttrs)

    if (edge.direction == "pc")  {     # specify appropriate direction!
      edgeRenderInfo(am.graph) <- list(arrowtail="open")
      edgeRenderInfo(am.graph) <- list(arrowhead="none")
#      edgeRenderInfo(am.graph) <- list(direction=NULL)# MESSES up!!! not needed.
    }
    edgeRenderInfo(am.graph) <- list(lwd=edge.lwd)

  #  if (plot) renderGraph(am.graph, attrs=attrs, nodeAttrs=nodeAttrs, edgeAttrs=edgeAttrs)
    if (plot) renderGraph(am.graph, ...)

#   } else {
#        am.graph <- layoutGraph(am.graph, attrs=attrs, nodeAttrs=nodeAttrs, edgeAttrs=edgeAttrs, ...)
        # the following does not work in R
#        edgeRenderInfo(am.graph)[["direction"]] <- "back"
        # hence
#        warning("edge.direction='pc' is not working with diamond shapes.")
#        edgeRenderInfo(am.graph) <- list(lwd=edge.lwd)
#        if (plot) renderGraph(am.graph,attrs=attrs, nodeAttrs=nodeAttrs, edgeAttrs=edgeAttrs)
#    }

    invisible(am.graph)
}  #EOF
